### PR TITLE
Enabled getApibyUUID, enabled APIs api for tenant mode by setting tenant flows

### PR DESCRIPTION
**Pull request by thilinicooray**
Added new property called artifactId for both backend and swagger API models
Modified impl methods for allowing get APIs by UUID
Set usernames and tenant domains inside started tenant flows to get tenant aware details from tenant registry spaces.

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/model/API.java [hunk 1]
```diff
@@ -28,6 +28,10 @@
 
     private APIIdentifier id;
 
+    // registry artifact id
+    // this id is provider's username independent
+    private String artifactId;
```
Comment (hevayo) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/model/API.java]: Hmmm , might be inconsistent with other models may be we can call it UUID in all the other models as well.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api/src/gen/java/org/wso2/carbon/apimgt/rest/api/dto/APIDTO.java [hunk 2]
```diff
@@ -11,6 +11,7 @@
 public class APIDTO  {
   
   private String name = null;
+  private String artifactId = null;
```
Comment (hevayo) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/gen/java/org/wso2/carbon/apimgt/rest/api/dto/APIDTO.java]: As for the model this should be ID.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/resources/api.yaml [hunk 3]
```diff
@@ -1393,9 +1393,13 @@ definitions:
       - name
       - context
       - version
+      - apiDefinition
     properties:
       name:
         type: string
+      artifactId:
```
Comment (hevayo) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/resources/api.yaml]: again this should be id not artifactId

File: components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApisApiServiceImpl.java [hunk 4]
```diff
@@ -108,8 +116,7 @@ public Response apisPost(APIDTO body,String contentType){
                 PrivilegedCarbonContext.endTenantFlow();
             }
         }
-
-        return Response.created(createdApiUri).entity(body).build();
+        return Response.created(createdApiUri).entity(createdApiDTO).build();
```
Comment (hevayo) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApisApiServiceImpl.java]: Goof find :)

File: components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/utils/mappings/APIMappingUtil.java [hunk 5]
```diff
@@ -228,6 +227,67 @@ public static APIDTO fromAPItoDTO(org.wso2.carbon.apimgt.api.model.API model) th
...
+                return APIDTO.SubscriptionAvailabilityEnum.all_tenants;
+            case APIConstants.SUBSCRIPTION_TO_SPECIFIC_TENANTS :
+                return APIDTO.SubscriptionAvailabilityEnum.specific_tenants;
+            default:
+                return null; // how to handle this?
+        }
+
+    }
+
+    private static String mapSubscriptionAvailabilityFromDTOtoAPI(
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/utils/mappings/APIMappingUtil.java]: Can we use single method coz functionality of abve method looks similar

File: components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/utils/mappings/APIMappingUtil.java [hunk 6]
```diff
@@ -228,6 +227,67 @@ public static APIDTO fromAPItoDTO(org.wso2.carbon.apimgt.api.model.API model) th
...
+            case RESTRICTED:
+                return APIConstants.API_RESTRICTED_VISIBILITY;
+            case CONTROLLED:
+                return APIConstants.API_CONTROLLED_VISIBILITY;
+            default:
+                return null; // how to handle this?
+        }
+    }
+    private static APIDTO.VisibilityEnum mapVisibilityFromAPItoDTO(String visibility) {
+        switch (visibility) { //public, private,controlled, restricted
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/utils/mappings/APIMappingUtil.java]: Check possibility of using single metthod

File: components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/utils/mappings/APIMappingUtil.java [hunk 7]
```diff
@@ -22,13 +22,7 @@
...
 import org.wso2.carbon.apimgt.api.APIManagementException;
 import org.wso2.carbon.apimgt.api.APIProvider;
-import org.wso2.carbon.apimgt.api.model.APIIdentifier;
-import org.wso2.carbon.apimgt.api.model.APIStatus;
-import org.wso2.carbon.apimgt.api.model.Documentation;
-import org.wso2.carbon.apimgt.api.model.DocumentationType;
-import org.wso2.carbon.apimgt.api.model.Scope;
-import org.wso2.carbon.apimgt.api.model.Tier;
-import org.wso2.carbon.apimgt.api.model.URITemplate;
+import org.wso2.carbon.apimgt.api.model.*;
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/utils/mappings/APIMappingUtil.java]: Avoid using wild cards

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIManager.java [hunk 8]
```diff
@@ -54,6 +54,14 @@
     /**
      * Returns details of an API
      *
+     * @param uuid UUID of the API's registry artifact
+     * @return An API object related to the given artifact id or null
+     * @throws APIManagementException if failed get API from APIIdentifier
+     */
+    public API getAPIbyUUID(String uuid) throws APIManagementException;
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIManager.java]: Leave blank line

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java [hunk 9]
```diff
@@ -331,6 +331,27 @@ public API getAPI(APIIdentifier identifier) throws APIManagementException {
...
+     * Get API by registry artifact id
+     *
+     * @param uuid  Registry artifact id
+     * @return API of the provided artifact id
+     * @throws APIManagementException
+     */
+    public API getAPIbyUUID(String uuid) throws APIManagementException {
+        try {
+            //registry of the current logged in user has been directly used
+            GenericArtifactManager artifactManager = APIUtil.getArtifactManager(registry, APIConstants.API_KEY);
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java]: is it need to load governance artifacts before retrieving artifact?

File: components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/utils/mappings/APIMappingUtil.java (hunk 7 above)
Comment (thilinicooray) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/utils/mappings/APIMappingUtil.java]: Fixed in PR #1091

File: components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/utils/mappings/APIMappingUtil.java (hunk 5 above)
Comment (thilinicooray) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/utils/mappings/APIMappingUtil.java]: Please consider that one method is mapping DTO to backend model while the other is mapping vice versa.
Method comments were added in PR #1091

File: components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/utils/mappings/APIMappingUtil.java (hunk 6 above)
Comment (thilinicooray) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/utils/mappings/APIMappingUtil.java]: Please consider that one method is mapping DTO to backend model while the other mapping vice versa.
Method comments were added in PR #1091

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIManager.java (hunk 8 above)
Comment (thilinicooray) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIManager.java]: Fixed in #1091 commit : 897ef52

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java (hunk 9 above)
Comment (thilinicooray) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java]: Please note that APIUtil. getArtifactManager() calls GovernanceUtils.loadGovernanceArtifacts((UserRegistry) registry) inside the method.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java [hunk 10]
```diff
@@ -331,6 +331,27 @@ public API getAPI(APIIdentifier identifier) throws APIManagementException {
...
+
+            GenericArtifact apiArtifact = artifactManager.getGenericArtifact(uuid);
+            return APIUtil.getAPIForPublishing(apiArtifact, registry);
+
+        } catch (RegistryException e) {
+            handleException("Failed to get API", e);
```
Comment (nuwand) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java]: handlerException method thrown an Exception right? If so the return statement below this line of code is never reachable. We can remove it.

Comment (thilinicooray) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java]: Fixed in PR #1280

File: components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApisApiServiceImpl.java [hunk 11]
```diff
@@ -190,28 +203,31 @@ public Response apisApiIdGet(String apiId,String accept,String ifNoneMatch,Strin
                 PrivilegedCarbonContext.endTenantFlow();
             }
         }
-
         return Response.ok().entity(apiToReturn).build();
     }
     @Override
     public Response apisApiIdPut(String apiId,APIDTO body,String contentType,String ifMatch,String ifUnmodifiedSince){
+
```
Comment (lafernando) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApisApiServiceImpl.java]: Remove extra space.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApisApiServiceImpl.java [hunk 12]
```diff
@@ -221,20 +237,21 @@ public Response apisApiIdPut(String apiId,APIDTO body,String contentType,String
...
         }
-
-        return Response.ok().entity(body).build();
+        return Response.ok().entity(updatedApiDTO).build();
     }
     @Override
     public Response apisApiIdDelete(String apiId,String ifMatch,String ifUnmodifiedSince){
         APIIdentifier apiIdentifier = APIMappingUtil.getAPIIdentifier(apiId);
         boolean isTenantFlowStarted = false;
         try{
```
Comment (lafernando) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApisApiServiceImpl.java]: Fix formatting.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApisApiServiceImpl.java (hunk 11 above)
Comment (thilinicooray) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApisApiServiceImpl.java]: Fixed in PR #1317

File: components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApisApiServiceImpl.java (hunk 12 above)
Comment (thilinicooray) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/impl/ApisApiServiceImpl.java]: Fixed in PR #1317

**Pull request merged**